h2o-core/src/main/java/water/util/IcedHashMapBase.java

```diff
@@ -28,13 +73,57 @@
   public Set<K> keySet()                                { return map().keySet(); }
   public Collection<V> values()                         { return map().values(); }
   public Set<Entry<K, V>> entrySet()                    { return map().entrySet(); }
-  public boolean equals(Object o)                       { return map().equals(o); }
-  public int hashCode()                                 { return map().hashCode(); }
+  @Override public boolean equals(Object o)             { return map().equals(o); }
+  @Override public int hashCode()                       { return map().hashCode(); }
+  @Override public String toString()                    { return map().toString(); }
+
+  private static final byte empty_map = -1;  // for future extensions, -1 becomes valid mode if for all enums there's an entry corresponding to their max allocated bits
+  static KeyType keyType(byte mode) { return KeyType.values()[mode & 0x3]; } // first 2 bits encode key type
+  static ValueType valueType(byte mode) { return ValueType.values()[mode>>>2 & 0xF];} // 3rd to 6th bit encodes value type
+  static ArrayType arrayType(byte mode) { return ArrayType.values()[mode>>>6 & 0x3]; } // 7th to 8th bit encodes array type
+
+  private static byte getMode(KeyType keyType, ValueType valueType, ArrayType arrayType) {
+    return (byte) ((arrayType.ordinal() << 6) | (valueType.ordinal() << 2) | (keyType.ordinal()));
+  }
+
+  private KeyType getKeyType(K key) {
+    assert key != null;
+    return Stream.of(KeyType.values())
+            .filter(t -> isValidKey(key, t))
+            .findFirst()
+            .orElseThrow(() -> new IllegalArgumentException("keys of type "+key.getClass().getTypeName()+" are not supported"));
+  }
+
+  private ValueType getValueType(V value) {
+    ArrayType arrayType = getArrayType(value);
+    return Stream.of(ValueType.values())
+            .filter(t -> isValidValue(value, t, arrayType))
```

Comment:
If we had a mapping from the combination of (arrayType, value) to the final ValueType, the time to lookup the final value could be shorter, instead of iterating through all the ValueTypes all the time. This is core functionality, that's why I'm thinking about decision trees speeding it all up a tiny bit.

Just thinking out loud.

Reply:
@Pscheidl It could be useful if someone decide to use similar logic for `IcedHashMapGeneric`. But for this implementation, the valueType is retrieved only once per object, so the difference will be negligible.

Reply:
Up to you to decide. Yes, it will be negligible.